documentation/projects/proposals/rekognition_data/20240530-implementation_plan_augment_catalog_with_rekognition_tags.md

```diff
@@ -0,0 +1,669 @@
+- **Average number of labels per image higher than 90**: 4.6629
+
+_For a full explanation on this exploration, see:
+[Analysis explanation](#analysis-explanation)_
+
+Based on the number of labels we would still be receiving with a confidence
+higher than 90, and that 0.9 is already our existing minimum standard, **we
+should retain 0.9 or 90% as our minimum label accuracy value** for inclusion in
+the catalog.
+
+This necessarily means that we will not be including a projected 62% of the
+labels which are available in the Rekognition dataset. Accuracy, as it directly
+relates to search relevancy, is more desirable here than completeness. We will
+retain the original Rekognition source data after ingesting the high-accuracy
+tags, and so if we decide to allow a lower accuracy threshold, we can always
+re-add the lower confidence values at a later time.
+
+## Step-by-step plan
+
+<!--
+List the ordered steps of the plan in the form of imperative-tone issue titles.
+The goal of this section is to give a high-level view of the order of implementation any relationships like
+blockages or other dependencies that exist between steps of the plan. Link each step to the step description
+in the following section.
+
+If special deployments are required between steps, explicitly note them here. Additionally, highlight key
+milestones like when a feature flag could be made available in a particular environment.
+-->
+
+In order to incorporate accomplish the goals of this plan, the following steps
+will need to be performed:
+
+1. [Determine which labels to exclude from Rekognition's label set](#determine-excluded-labels)
+2. [Remove the above excluded labels and tags below threshold from existing Clarifai tags](#filter-clarifai-tags)
+3. [Generate and insert the new Rekognition tags](#insert-new-rekognition-tags)
+
+## Step details
+
+<!--
+Describe all of the implementation steps listed in the "step-by-step plan" in detail.
+
+For each step description, ensure the heading includes an obvious reference to the step as described in the
+"step-by-step plan" section above.
+-->
+
+```{note}
+Some of the steps listed below have some cross-over with functionality defined
+in/required by the
+[data normalization project](/projects/proposals/data_normalization/20240227-implementation_plan_catalog_data_cleaning.md)
+(#430). Where possible, existing issues will be referenced and possible duplicated
+effort will be identified.
+```
+
+### Determine excluded labels
+
+This will involve a manual process of looking through each of the [available
+labels for Rekognition][aws_rekognition_labels] and seeing if they match any of
+the criteria to be filtered. The excluded labels should then be saved in an
+accessible location, either on S3 or within the
+[sensitive terms repository](https://github.com/WordPress/openverse-sensitive-terms)
+as a new file. Consent & approval should be sought from two other maintainers on
+the accuracy of the exclusion list prior to publishing.
```

Comment:
Some observations that might inform how much detail we put into this part of the plan, and how we anticipate this work of determining which labels to use will go:

- There are 3082 labels in the "all labels" file of the linked zip. Will someone really look through each line of this file and determine which to include or exclude? How will this work be reviewed?
- I wasn't able to find any of the following terms in that list: muslim, christian, jesus, driver, steward, host, pilot, salesman (or saleswoman), homosexual, gay, transgender (or related terms).
- None of the listed professions (terms in the profession category) are gendered, but almost all of them are focused on urban, non-rural professions (except some are historical like knight, samurai, ninja, and arguably "pirate" depending on the context). That's just an interesting _different_ bias than I originally anticipated. The only one related to working outside is "gardner". The rest are performance related (magician, performer, ballerina, pianist) or things like surgeon, executive, teacher, veterenarian, soldier. Sniper, is weirdly included, which would be an interesting one to do analysis on and cross-compare our existing meta data to see where and when that label gets applied by Rekognition.
- Similarly, it includes "Monk", but not "Nun". Another interesting bias, and I wonder why they excluded nun. My guess is that it's difficult for computers to reliably distinguish between different types of head coverings. But do they only include Trappist-type monks? Or also Buddhist monks? Are all the monks medieval European monks? Interesting one!
- The rekognition label set is actually a lot _smaller_ than I thought, and way less controversial than I thought. I think there is less chance of the rekognition labels causing issues related to the two groups of label categories above, and in fact there are only two labels that I think are easy to argue for exclusion: man, boy, female, male, lady, woman, girl. On the other hand, funny enough, martial status is probably the second easiest to argue for exclusion, due to the almost inherently gendered content of these terms. Rekognition includes bride, bridegroom, bridesmaid. It does not include wife, husband, or partner.

I do not think we should rely on Rekognition's labels to inform our analysis of Clarifai's labels. We probably need to pull all the Clarifai labels into a list and evaluate them separately. I cannot find any specific list of labels on Clarifai's documentation (I spent only a little less than 10 minutes clicking around so I might have missed it), and my impression is that Clarifai has lots of different models, not just one (like Rekognition) and that even the source of "Clarifai" is maybe not that useful. It's be nice if we had the actual model they came from.

Can we pull the list of all Clarifai labels in the catalogue for evaluation? Is there a specific provider that has the Clarifai labels or is it spread across multiple providers? If it's only from one provider, then I wonder if we can reach out and ask them if they created those labels and what model they used if so. If it's from more than one, then probably that data is historical and from some project done during CC days? In which case, we should reach out to folks involved back then to see if they remember how those labels were acquired/generated to give us more information.

Reply:
> There are 3082 labels in the "all labels" file of the linked zip. Will someone really look through each line of this file and determine which to include or exclude? How will this work be reviewed?

Yes, that was my intention and I can make this explicit in the IP if you'd like. Do you think it would be better to have two people go through this process and share lists at the end?

> We probably need to pull all the Clarifai labels into a list and evaluate them separately. I cannot find any specific list of labels on Clarifai's documentation (I spent only a little less than 10 minutes clicking around so I might have missed it), and my impression is that Clarifai has lots of different models, not just one (like Rekognition) and that even the source of "Clarifai" is maybe not that useful. It's be nice if we had the actual model they came from.

This was my experience as well; I looked for a list but was unable to find one. I like the idea of doing analysis on the labels we already have, potentially creating a list in a similar manner to the Rekognition labels provided and performing the same label selection there.

Just noting for myself: because of the separation between Rekognition and Clarifai, I don't think any of the Clarifai steps (since the data is already present!) need to prevent us from moving forward on the Rekognition data incorporation end (barring the other discussions happening in this IP).

Reply:
> Yes, that was my intention and I can make this explicit in the IP if you'd like. Do you think it would be better to have two people go through this process and share lists at the end?

That sounds good :+1: 

> Just noting for myself: because of the separation between Rekognition and Clarifai, I don't think any of the Clarifai steps (since the data is already present!) need to prevent us from moving forward on the Rekognition data incorporation end (barring the other discussions happening in this IP).

Agreed. We could exclude Clarifai tags from the searched tags when we do this work to exclude the tags from search (but keeping them in the catalog). And then add two more issues to this project, that essentially are just the same tasks for checking the Rekognition labels but doing so for Clarifai. So...

1. Exclude Clarifai tags from search by filtering them out during data refresh
2. Pull the Clarifai tags and decide which ones to use
3. Filter in the Clarifai tags we want to use during data refresh (which is the same thing we're already doing for Rekognition, so this should just be a new case in existing code, rather than an entirely new feature)

It does add new work, but I think we should track it and create the issues as a result of this IP, but if you want to exclude it from the project's "shipped" status to prevent adding that work to the scope of this project, that sounds fine to me. But we should track it and we might as well base that process on what we're doing for Rekognition, so that we're being consistent and following the same approach.

Reply:
+1 for having two people go through the process and share lists. I was going to recommend that reviewers of the excluded label list go through the entire source list when reviewing, since you would have to in order to make sure that all labels that needed to be excluded were properly identified. But doing the filtering independently and comparing afterward is an even better idea IMO!

> Just noting for myself: because of the separation between Rekognition and Clarifai, I don't think any of the Clarifai steps (since the data is already present!) need to prevent us from moving forward on the Rekognition data incorporation end (barring the other discussions happening in this IP).

Also agreed. I like the idea of doing analysis on the labels we already have. Shame we can't find a documented list :/

Reply:
Just want to note that https://github.com/WordPress/openverse/pull/4417#discussion_r1626779131 would have some implications for how to handle the Clarifai data. If we treat the Rekognition S3 -> Cat. DB ETL as a repeatable process and do the filtering there, then we won't have a mechanism for filtering Clarifai data. We'd need to similarly pull the Clarifai data out of the media tables (into S3? Into another table?) and then have an ETL process for loading in the Clarifai tags we want for each record based on filtering logic. Or else, we need to design the filtering of Rekognition tags such that it's directly usable for Clarifai tags, which means filtering it at the Cat. DB -> API DB/ES stage, which has it's own complications we're discussing in that other thread.

I don't want to segment the conversation from that thread, it's probably better to have it all in one place, and I've noted these issues there too, but I want to add that context to this one because it slightly complicates the 3 new tasks I suggested in my previous comment here if we filter Rekognition tags before inserting them into the Cat. DB because that's not a directly reusable/extendable process for the Clarifai tags as they are now.